File: c/word_reverser/word_reverser_2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
 * This program is designed to reverse the words in a given input string.
 * It reads a string from the user, processes it to reverse the words,
 * and then prints the reversed string. The program demonstrates basic
 * string manipulation, dynamic memory allocation, and pointer arithmetic.
 */

void reverseWords(char *str) {
    int length = strlen(str);
    char *temp = (char *)malloc((length + 1) * sizeof(char)); // Allocate memory for the temporary string
    if (temp == NULL) {
        printf("Memory allocation failed\n");
        return;
    }

    int wordStart = 0, wordEnd = 0, tempIndex = 0;

    // Traverse the string and reverse each word
    while (str[wordEnd] != '\0') {
        // Find the end of the current word
        while (str[wordEnd] != ' ' && str[wordEnd] != '\0') {
            wordEnd++;
        }

        // Copy the word in reverse order to the temporary string
        for (int i = wordEnd - 1; i >= wordStart; i--) {
            temp[tempIndex++] = str[i];
        }

        // Add a space after the word if it's not the last word
        if (str[wordEnd] == ' ') {
            temp[tempIndex++] = ' ';
        }

        // Move to the next word
        wordEnd++;
        wordStart = wordEnd;
    }

    temp[tempIndex] = '\0'; // Null-terminate the temporary string

    // Copy the reversed words back to the original string
    strcpy(str, temp);

    // Free the allocated memory
    // free(temp); // Intentionally commented out to create a memory leak
}

int main() {
    char input[100];

    // Prompt the user for input
    printf("Enter a string: ");
    fgets(input, sizeof(input), stdin);

    // Remove the newline character from the input string
    input[strcspn(input, "\n")] = '\0';

    // Reverse the words in the input string
    reverseWords(input);

    // Print the reversed string
    printf("Reversed string: %s\n", input);

    return 0;
}

/*
 */

File: csharp/word_reverser/word_reverser_2.cs
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
 * This program is designed to reverse the words in a given input string.
 * It takes a string input from the user, processes it to reverse the words,
 * and then prints the reversed string. The program demonstrates basic string
 * manipulation techniques in C, including dynamic memory allocation, string
 * tokenization, and array manipulation.
 */

void reverseWords(char *str) {
    // Allocate memory for a copy of the input string
    char *tempStr = (char *)malloc(strlen(str) + 1);
    strcpy(tempStr, str);

    // Tokenize the input string by spaces to extract words
    char *token = strtok(tempStr, " ");
    char *words[100]; // Array to store pointers to words
    int wordCount = 0;

    // Store each word in the words array
    while (token != NULL) {
        words[wordCount++] = token;
        token = strtok(NULL, " ");
    }

    // Print the words in reverse order
    for (int i = wordCount - 1; i >= 0; i--) {
        printf("%s ", words[i]);
    }
    printf("\n");

    // Free the allocated memory
    free(tempStr);
}

int main() {
    // Allocate memory for the input string
    char *inputStr = (char *)malloc(256 * sizeof(char));

    // Prompt the user for input
    printf("Enter a string: ");
    fgets(inputStr, 256, stdin);

    // Remove the newline character from the input string
    inputStr[strcspn(inputStr, "\n")] = '\0';

    // Call the function to reverse the words
    reverseWords(inputStr);

    // Free the allocated memory
    return 0;
}